### ISSUE TITLE
clarification of 'split' event

### ISSUE THREAD

**Issue by dominictarr** [dodo/node-bufferstream#4]
I'm not sure from your readme what the split event emits?

does it emit the buffer up to the just before the token starts?

also, does it join buffers together if they do not already end in a token?

in other words, can I use this to rechunk a stream so that the chunks always break on newlines, for example?

**Comment by dodo** [comment]
> I'm not sure from your readme what the split event emits?
> you can specify more than one split token .. so it's emitted whenever
> a token is found.
> 
> does it emit the buffer up to the just before the token starts?
> yes.
> 
> also, does it join buffers together if they do not already end in a token?
> when size is `flexible` it joins everything together what it gets to
> one buffer (accessible through `stream.buffer` or
> `stream.getBuffer()`)
> whenever it gets data, it will try to find all tokens
> 
> in other words, can I use this to rechunk a stream so that the chunks always break on newlines, for example?
> yes.
> 
> ``` javascript
> stream = new BufferStream({size:'flexible'});
> stream.split('\n', function (line) { // line doesn't have a '\n' anymore
>     stream.emit('data', line); // Buffer.isBuffer(line) === true
> });
> ```

if you have more question, or does anything concerns you about the
api, please point them out.

anyways, i'll update the readme :)

**Comment by dominictarr** [comment]
that is still a little confusing, but I assume that if the raw stream is like this:

```
"I am a raw stream\n he", "ar me R", "OAR\n", "!!!!!1!!!" 
```

if you are splitting on '\n', the split event will give you:

   "I am a raw stream", "hear me ROAR", "!!!!!1!!!"
- excluding the tokens,
- still emitting the chunks before the first token and after the last token?

is that correct?

the only thing that I think you should add is to go 

```
new BufferStream({size: 'flexible', split: token})
```

and then have the BufferStream emit the "split" events as "data" events.  
this would mean that you could use pipe to send it to another stream.

**Comment by dodo** [comment]
not quite,  "!!!!!1!!!" would be stuck in the buffer because there is no "\n" after it to cut it off.

btw thanks for the idea ... plz moar of that ;)

**Comment by dominictarr** [comment]
ah, it might be a good idea to also still emit the last thing.

that is usually what you'd want if you are splitting by newlines...

maybe add an {inclusive: boolean} option, and default it to true?

**Comment by dominictarr** [comment]
another important nuance:

what happens if the stream has split tokens immediately following each other "data\n\ndada"?

in http, a double newline is used to signify the end of the header section.

**Comment by dodo** [comment]
just discovered, that this allready works :D

so it will emit "!!!!!1!!!" on end.

**Comment by dominictarr** [comment]
awesome!

**Comment by dodo** [comment]
``` javascript
stream = new (require('bufferstream'))({size:'flexible', split:'\n'});
stream.on('data', function (line) {console.log(line.length, line.toString())});
stream.write("data\n\ndada");
stream.end();
```

results in:

```
4 'data'
0 ''
4 'dada'
```

**Comment by dominictarr** [comment]
that is good.

**Comment by dodo** [comment]
btw .. i'm allready cutting off a http header from a cgi script with bufferstream https://codetu.be/team/codetube/blob/master/src/git-http-backend.coffee#L105

**Comment by asnowfix** [comment]
It it possible to use this without setting-up the 'split' token? I would like to use at as a Stream interface over a Buffer (or a series of Buffers...), without having to wait for any specific token ('data' event is emitted as soon as there is data in the buffer). Did I knock the wrong door?

**Comment by dodo** [comment]
yes it is possible. just don't use split :)

**Comment by asnowfix** [comment]
So this is a bug, because no 'data' event is ever sent without a 'split' set, or at least this is what I have found so far...

**Comment by dodo** [comment]
where did you found it? have you actually tried it?

**Comment by asnowfix** [comment]
I would have bothered reporting an issue without having actually tried it. believe it or not ;-)

I am now using https://github.com/samcday/node-stream-buffer, which does the trick, although it is less sophisticated/feature-ruch compared to node-bufferstream.
